ARCs/arc-0018.md

```diff
@@ -0,0 +1,348 @@
+
+The Freeze role MUST be set to the Application Address if `FreezeAddr != ZeroAddress`, else set to `ZeroAddress`.
+
+> If the creator wants to allow an ASA to be Royalty Free after some conditions are met, it should be set to the Application Address 
+
+The Manager role MUST be set to the Application Address if `ManagerAddr != ZeroAddress`, else set to `ZeroAddress`.
+
+> If the creator wants to update the Freeze role, this should be set to the application address
+
+The Reserve role MAY be set to anything.
+
+#### Asset Parameters
+
+The `DefaultFrozen` MUST be set to true.
+
+
+### Third Party Marketplace
+
+In order to support secondary sales on external markets this spec was designed such that the Royalty Asset may be listed without transferring it from the current owners account.  A Marketplace may call the transfer request as long as the address initiating the transfer has been set as the `authorized_address` through the [offer](#offer) method in some previous transaction by the current owner.
+
+
+### ABI Methods
+
+#### Required
+
+
+##### SetPolicy
+    
+    set_policy(royalty_asset: asset, royalty_receiver: address, royalty_share: uint64, allowed_asset_1: asset, allowed_asset_2: asset, allowed_asset_3: asset, allowed_asset_4: asset)
+
+Sets the Royalty Policy for a given asset. 
+
+The `royalty_asset` is the ASA id of the asset the policy is set for.  If the `royalty_asset` is 0, the policy is defined for _any_ asset that has its freeze and clawback set to the application address. 
+
+The `royalty_receiver` is the address of the account that should receive a partial share of the payment or asset transfer for any [transfer](#transfer) of the `royalty_asset` 
+
+The `royalty_share` is the share of the payment that should be sent to the `royalty_receiver`.  It is specified in basis points (e.g., 1% is 100).  
+
+The `allowed_assets_n` arguments represents the ASA ids that are acceptable for payment. The contract logic MUST opt into the set of assets specified in order to accept the payments as part of a transfer.
+
+> TODO: since assets cant be in dynamic arrays, this bit is annoying. We could define N selectors (up to 8 for assets) and dispatch them all to the same set of logic? for now just hardcoded to 4 and caller expected to pass 0s if they don't need all 4.  As of https://github.com/algorandfoundation/ARCs/pull/77, this limitation has been changed. Update spec accordingly once SDK support is available.
+
+>The Royalty Collector SHOULD be opted into the full set of assets contained in this list but MAY NOT be checked ahead of time. In the case that an account is not opted into an asset, any transfers where payment is specified for that asset will fail until the account opts into the asset.
+
+
+##### Transfer: 
+
+    transfer(royalty_asset: asset, from: account, to: account, royalty_collector: account, royalty_asset_amount: uint64, payment: transaction)
```

Comment:
What happens of not all the offered amount is used?
Can a second transfer take place?
Or does it rescind the offer?

Reply:
In the current reference implementation the offered amount is updated to reflect the new offered amount but maybe we should be more explicit about it? It may also make sense to just wipe it completely.

I'm not sure what the correct behavior is, wdyt?

Reply:
I've updated the document to describe how the implementer might consider updating the offer after a transfer takes place.